features/crc/soft/FastCRC.h

```diff
@@ -0,0 +1,216 @@
+                _polynomial = 0x07;            // x8+x2+x+1
+                _width = 8;
+                break;
+
+            case CRC_16BIT_SD:
+                _polynomial = 0x1021;          // x16+x12+x5+1
+                _width = 16;
+                break;
+
+            case CRC_16BIT_CCITT:
+                _polynomial = 0x1021;          // x16+x12+x5+1
+                _width = 16;
+                _inital_value = ~(0x0);
+                break;
+
+            case CRC_16BIT_IBM:
+                _polynomial = 0x8005;          // x16+x15+x2+1
+                _width = 16;
+                _reflect_data = true;
+                _reflect_remainder = true;
+                break;
+
+            case CRC_32BIT:
+                _polynomial = 0x04C11DB7;     // x32+x26+x23+x22+x16+x12+x11+x10+x8+x7+x5+x4+x2+x+1
+                _width = 32;
+                _inital_value = ~(0x0);
+                _final_xor  = ~(0x0);
+                _reflect_data = true;
+                _reflect_remainder = true;
+                break;
+        }
+        MBED_ASSERT(sizeof(_inital_value) == CRC_GET_DATA_SIZE(_width));
+    }
+
+    virtual ~FastCRC() { }
+
+    /** Initialize a CRC module, generate CRC static table
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t init()
+    {
+        T remainder;
+        uint32_t polynomial = (_width < 8) ? (_polynomial << (8 - _width)) : _polynomial;
+
+        for (int32_t dividend = 0; dividend < 256; ++dividend) {
+            if ( _width < 8) {
+                remainder = dividend;
+            } else {
+                remainder = dividend << (_width - 8);
+            }
+            for (uint8_t bit = 8; bit > 0; --bit) {
+                if (remainder & CRC_TOP_BIT(_width)) {
+                    remainder = (remainder << 1) ^ polynomial;
+                } else {
+                    remainder = (remainder << 1);
+                }
+            }
+
+            /*
+             * Store the result into the table.
+             */
+            _crc_table[dividend] = remainder;
+        }
+        return 0;
+    }
+
+    /** Deinitialize a CRC module
+     *
+     * @return  0 on success, negative error code on failure
+     */
+    virtual int32_t deinit()
+    {
+        return 0;
+    }
+
+    /** Get CRC polynomial
+     *
+     * Get the current CRC polynomial
+     *
+     * @return  Polynomial value
+     */
+    virtual uint32_t get_polynomial(void) const
+    {
+        return _polynomial;
+    }
+
+    /** Compute CRC for the data input as per the polynomial set
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC
+     *  @return  0  on success or a negative error code on failure
+     *  @note CRC calculations are done byte wise.
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, uint32_t *crc)
+    {
+        MBED_ASSERT(crc != NULL);
```

Comment:
Now that these aren't templates, could the implementations be moved back into C++ files?

Reply:
FastCRC implementation is template based